hardware_interface/include/hardware_interface/sensor.hpp

```diff
@@ -0,0 +1,204 @@
+   * have the same length; return_type::INTERFACE_NOT_FOUND if one of provided interfaces is not
+   * defined for the sensor; return_type::OK otherwise.
+   */
+  HARDWARE_INTERFACE_EXPORT
+  return_type get_state(
+    std::vector<double> & state,
+    const std::vector<std::string> & interfaces) const
+  {
+    if (interfaces.size() == 0) {
+      return return_type::INTERFACE_NOT_PROVIDED;
+    }
+    return_type ret = return_type::OK;
+    bool found;
+
+    for (const auto & interface : interfaces) {
+      found = false;
+      for (uint i = 0; i < info_.state_interfaces.size(); i++) {
+        if (!interface.compare(info_.state_interfaces[i])) {
+          state.push_back(states_[i]);
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        ret = return_type::INTERFACE_NOT_FOUND;
+        state.clear();
+        break;
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * \brief Get complete state list from the sensor. This function is used by the controller to get
+   * complete actual state of the hardware. The state values have the same order as interfaces which
+   * can be recived by get_state_interfaces() function.
+   *
+   * \param state list of doubles with states of the hardware.
+   */
+  HARDWARE_INTERFACE_EXPORT
+  void get_complete_state(std::vector<double> & state) const
+  {
+    state.clear();
+    for (const auto & internal_state : states_) {
+      state.push_back(internal_state);
+    }
+  }
+
+  /**
+   * \brief Set state list for the sensor. This function is used by the hardware to set its actual
+   * state. The parameters state, and interfaces have the same order and number of elements. Using
+   * the interfaces list, the hardware can choose which values to set.
+   *
+   * \param state list of doubles with states of the hardware.
+   * \param interfaces list of interfaces on which states have to be provided.
+   * \return return_type::INTERFACE_VALUE_SIZE_NOT_EQUAL if state and interfaces arguments do not
+   * have the same length; return_type::INTERFACE_NOT_FOUND if one of provided interfaces is not
+   * defined for the sensor; return_type::OK otherwise.
+   */
+  HARDWARE_INTERFACE_EXPORT
+  return_type set_state(
+    const std::vector<double> & state,
+    const std::vector<std::string> & interfaces)
+  {
+    if (state.size() != interfaces.size()) {
+      return return_type::INTERFACE_VALUE_SIZE_NOT_EQUAL;
+    }
+    return_type ret = return_type::OK;
+    bool found;
+
+    for (uint i = 0; i < info_.state_interfaces.size(); i++) {
+      found = false;
+      for (uint j = 0; j < info_.state_interfaces.size(); j++) {
+        if (!interfaces[i].compare(info_.state_interfaces[j])) {
+          states_[j] = state[i];
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        ret = return_type::INTERFACE_NOT_FOUND;
+        break;
+      }
+    }
```

Comment:
as per comment above, this logic should actually be refactored in a free function. That way, there's no code duplication.